### PR TITLE
Coffee Shop 

### PR DESCRIPTION
**Pull request by elimelec**

### CODE REVIEW

File: app/tests/LocationTest.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import org.junit.jupiter.api.Test
+import kotlin.test.assertEquals
+
+class LocationTest {
+    @Test
+    fun `test location`() {
+        val location = Location(12.51, -50.12)
+        assertEquals(12.51, location.x)
```
Comment (alexandru-calinoiu) [inline app/tests/LocationTest.kt:8]: If `Location` is a data class, it has some useful properties that will allow you to write only one assert.

File: app/src/Location.kt [hunk 2]
```diff
@@ -0,0 +1 @@
+data class Location(val x: Double, val y: Double)
```
Comment (alexandru-calinoiu) [inline app/src/Location.kt]: Not implying that you made a mistake (this applies to all my comments by the way) but why choose `Double` over a `Float`?

File: app/tests/CoffeeShopTest.kt [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import org.junit.jupiter.api.Test
+import kotlin.test.assertEquals
+
+class CoffeeShopTest {
+    @Test
+    fun `test coffee shop`() {
+        val shop = CoffeeShop("Starbucks", Location(21.51, -74.58))
+        assertEquals("Starbucks", shop.name)
+//        assertEquals(21.51, shop.x)
```
Comment (alexandru-calinoiu) [inline app/tests/CoffeeShopTest.kt]: This comments are not very usefull

File: app/tests/CoffeeShopTest.kt [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+//        assertEquals(-74.58, shop.y)
+        assertEquals(21.51, shop.location.x)
+        assertEquals(-74.58, shop.location.y)
+    }
+
+    fun `test coffee shop without using location`() {
```
Comment (alexandru-calinoiu) [inline app/tests/CoffeeShopTest.kt:14]: Good point, but still over work, you could have simply choose one path and if the need had arrisen write this second constructor.

File: app/tests/CoffeeShopTest.kt [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        val shop = CoffeeShop("Starbucks", 21.51, -74.58)
+        assertEquals("Starbucks", shop.name)
+//        assertEquals(21.51, shop.x)
+//        assertEquals(-74.58, shop.y)
+        assertEquals(21.51, shop.location.x)
```
Comment (alexandru-calinoiu) [inline app/tests/CoffeeShopTest.kt]: Accessing `x` and `y` like this makes me thing of the [feature envy code smell](https://duckduckgo.com/?q=feature+envy&t=ffab&atb=v129-2&ia=web)

File: app/tests/LocationTest.kt [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+        assertEquals(-50.12, location.y)
```
Comment (alexandru-calinoiu) [inline app/tests/LocationTest.kt:9]: I will argue that `x` and `y` are not great names for locations, why not go with `longitude` and `latitude` ?

File: app/tests/DistanceTest.kt [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import org.junit.jupiter.api.Test
+import kotlin.test.assertEquals
+
+class DistanceTest {
+    @Test
+    fun `test distance from examples`() {
```
Comment (alexandru-calinoiu) [inline app/tests/DistanceTest.kt]: To much of a test, a single example would have sufficed. And then consider doing a test for some edge case if any

File: app/tests/DistanceTest.kt [hunk 8]
```diff
@@ -2,19 +2,31 @@ import org.junit.jupiter.api.Test
 import kotlin.test.assertEquals
 
 class DistanceTest {
+
     @Test
-    fun `test distance from examples`() {
+    fun `test distance from examples (user - seattle2)`() {
```
Comment (alexandru-calinoiu) [inline app/tests/DistanceTest.kt:9]: Like the fact that you have 1 assert per test :+1: 

Still don't think is a good idea to add this many tests at this point.

File: app/src/Location.kt [hunk 9]
```diff
@@ -1 +1,10 @@
+import kotlin.math.abs
+
 data class Location(val x: Double, val y: Double)
+
+fun distance(location1: Location, location2: Location): Double {
+    var (x1, y1) = location1
```
Comment (alexandru-calinoiu) [inline app/src/Location.kt]: Destructuring the `Location` class very cool.

local `x`'s and `y`'s should be made immutable

File: app/src/CoffeeShop.kt [hunk 10]
```diff
@@ -13,3 +13,8 @@ fun String.toCoffeeShop(): CoffeeShop {
         throw IllegalArgumentException("row is invalid csv: $this")
     }
 }
+
+fun Array<String>.toCoffeeShops(): Array<CoffeeShop> {
```
Comment (alexandru-calinoiu) [inline app/src/CoffeeShop.kt]: Why choose an `Array` over a `List`?

File: app/src/CoffeeShop.kt [hunk 11]
```diff
@@ -14,7 +14,7 @@ fun String.toCoffeeShop(): CoffeeShop {
     }
 }
 
-fun Array<String>.toCoffeeShops(): Array<CoffeeShop> {
-    return this.map { it.toCoffeeShop() }.toTypedArray()
+fun List<String>.toCoffeeShops(): List<CoffeeShop> {
```
Comment (alexandru-calinoiu) [inline app/src/CoffeeShop.kt]: Ha :) Why choose a `List` over an `Array`?

File: app/tests/CoffeeShopWithDistanceTest.kt [hunk 12]
```diff
@@ -0,0 +1,10 @@
+import org.junit.jupiter.api.Test
+
+class CoffeeShopWithDistanceTest {
+    @Test
+    fun `test coffee shop with distance instance`() {
```
Comment (alexandru-calinoiu) [inline app/tests/CoffeeShopWithDistanceTest.kt:11]: Not really sure what this test was for initially, maybe I will understand later

File: app/tests/LocationTest.kt [hunk 13]
```diff
@@ -8,4 +8,11 @@ class LocationTest {
         assertEquals(12.51, location.x)
         assertEquals(-50.12, location.y)
     }
+
+    @Test
+    fun `test location equality`() {
+        val location1 = Location(12.51, -50.12)
+        val location2 = Location(12.51, -50.12)
+        assert(location1 == location2)
```
Comment (alexandru-calinoiu) [inline app/tests/LocationTest.kt]: You wrote a test for a kotlin feature, I think we can leave the kotlin developers write tests for their own features.

File: app/src/CoffeeShopWithDistance.kt [hunk 14]
```diff
@@ -1,2 +1,5 @@
 class CoffeeShopWithDistance(val shop: CoffeeShop, val location: Location) {
-}
\ No newline at end of file
+    val distance: Double by lazy {
```
Comment (alexandru-calinoiu) [inline app/src/CoffeeShopWithDistance.kt]: A `distance` class has a `distance` property intern.

File: app/tests/CoffeeShopWithDistanceTest.kt [hunk 15]
```diff
@@ -28,4 +28,11 @@ class CoffeeShopWithDistanceTest {
         val other = CoffeeShopWithDistance(shop, location)
         assert(distance == other)
     }
+
+    @Test
+    fun `test inequality`() {
+        val otherLocation = Location(20.12, 50.8)
+        val other = CoffeeShopWithDistance(shop, otherLocation)
+        assert(distance != other)
```
Comment (alexandru-calinoiu) [inline app/tests/CoffeeShopWithDistanceTest.kt]: I will argue that this test is somehow redundant.

File: app/tests/CoffeeShopWithDistanceTest.kt [hunk 16]
```diff
@@ -36,6 +36,7 @@ class CoffeeShopWithDistanceTest {
         assert(distance != other)
     }
 
+    @Test
```
Comment (alexandru-calinoiu) [inline app/tests/CoffeeShopWithDistanceTest.kt]: Obviously was not run before :)

File: app/tests/CoffeeShopAppTest.kt [hunk 17]
```diff
@@ -0,0 +1,11 @@
+import org.junit.jupiter.api.Test
+
+class CoffeeShopAppTest {
+    @Test
+    fun `test coffee shop app instance`() {
+        val app = CoffeShopApp(47.6, -122.4, "coffee_shop.csv")
+        app.x = 47.6
```
Comment (alexandru-calinoiu) [inline app/tests/CoffeeShopAppTest.kt]: Not sure why you made variables writable.

File: app/src/CoffeeShopApp.kt [hunk 18]
```diff
@@ -1,6 +1,6 @@
 class CoffeeShopApp (val x: Double, val y: Double, val filename: String) {
     fun getNearestShops(): List<CoffeeShopWithDistance> {
-        return listOf<CoffeeShopWithDistance>()
+        return listOf()
```
Comment (alexandru-calinoiu) [inline app/src/CoffeeShopApp.kt]: You can shorted it even more:

```
     fun getNearestShops(): List<CoffeeShopWithDistance> = listOf()
```
Or better yet try to use `emptyList()`

File: app/tests/CoffeeShopAppTest.kt [hunk 19]
```diff
@@ -24,6 +24,8 @@ class CoffeeShopAppTest {
     fun `test console output with example`() {
         val app = CoffeeShopApp(47.6, -122.4, "coffee_shops.csv")
         val output  = app.getOutput()
-        assertEquals("Starbucks Seattle2,0.0645\nStarbucks Seattle,0.0861\nStarbucks SF,10.0793", output)
+
+        val ls = System.lineSeparator()
+        assertEquals("Starbucks Seattle2,0.0645${ls}Starbucks Seattle,0.0861${ls}Starbucks SF,10.0793", output)
```
Comment (alexandru-calinoiu) [inline app/tests/CoffeeShopAppTest.kt:31]: :+1: did not knew about it.

File: app/src/CoffeeShopApp.kt [hunk 20]
```diff
@@ -0,0 +1,33 @@
+import java.io.File
+import java.io.FileNotFoundException
+
+class CoffeeShopApp (val x: Double, val y: Double, val filename: String) {
+    fun getNearestShops(): List<CoffeeShopWithDistance> {
```
Comment (alexandru-calinoiu) [inline app/src/CoffeeShopApp.kt]: This does not make sens to be public, you should limit the public surface, read more about it `Indecent Exposure` code smell.

File: app/src/CoffeeShopApp.kt [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+        catch (e: FileNotFoundException) {
+            return "$filename was not found."
+        }
+        catch (e: CouldNotParseCsvRowException) {
+            return "$filename is not a valid file."
+        }
+    }
+}
+
+fun readAllLinesAsCsvRows(filename: String): List<String> {
```
Comment (alexandru-calinoiu) [inline app/src/CoffeeShopApp.kt]: Again this method should not be public.

**Review by alexandru-calinoiu — COMMENTED** [review]
This is a decent implementation.

But the goal of the tdd was largely missed, you basically knew the implementation from the beginning and started with testing the details.

To get a better feel of how this can be properly done, you can check out this video https://www.youtube.com/watch?v=qkblc5WRn-U

Another nice touch would have been a description for the pr to  help the reviewer with setting up the environment.

Please review the comments and response when appropriate (they where done in the order of the commits and some of them migt make sens)

File: app/src/CoffeeShopApp.kt [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+            }
+            else {
+                return shops.joinToString(separator = System.lineSeparator())
+            }
+
+        }
+        catch (e: FileNotFoundException) {
+            return "$filename was not found."
+        }
+        catch (e: CouldNotParseCsvRowException) {
```
Comment (alexandru-calinoiu) [inline app/src/CoffeeShopApp.kt:25]: Can you think of some other ways to have resilience but have a nicer user experience?

File: app/src/Location.kt (hunk 2 above)
Comment (elimelec) [reply app/src/Location.kt]: I use `Double` by default since it has more precision and it's the default literal interpretation in most languages and never think about it if the need doesn't arise.

File: app/src/CoffeeShop.kt (hunk 10 above)
Comment (elimelec) [reply app/src/CoffeeShop.kt]: `Array` happened to be the first class I thought about when I wrote the test. I just needed a list like data type.

File: app/src/CoffeeShop.kt (hunk 11 above)
Comment (elimelec) [reply app/src/CoffeeShop.kt]: When I actually wrote code I noticed that stream functions return `List`, there was no point to always convert the `List` to `Array`, or to use different types across the app when one can handle everything just fine.

File: app/tests/LocationTest.kt (hunk 13 above)
Comment (elimelec) [reply app/tests/LocationTest.kt]: You're right. I'll remove them.

File: app/tests/CoffeeShopAppTest.kt (hunk 17 above)
Comment (elimelec) [reply app/tests/CoffeeShopAppTest.kt]: It was mistake, I wanted to write `assert(47.6, app.x)`... 😬

File: app/src/CoffeeShopApp.kt (hunk 18 above)
Comment (elimelec) [reply app/src/CoffeeShopApp.kt]: Nice, didn't know this.

File: app/src/CoffeeShopApp.kt (hunk 20 above)
Comment (elimelec) [reply app/src/CoffeeShopApp.kt]: I forgot about this. Will update the code.

File: app/tests/LocationTest.kt (hunk 1 above)
Comment (elimelec) [reply app/tests/LocationTest.kt:8]: Are you thinking about `@ParameterizedTest`?

File: app/src/CoffeeShopWithDistance.kt (hunk 14 above)
Comment (elimelec) [reply app/src/CoffeeShopWithDistance.kt]: I don't understand what you want to say.

File: app/tests/LocationTest.kt (hunk 6 above)
Comment (elimelec) [reply app/tests/LocationTest.kt:9]: I don't think `x` and `y` are bad given that the specification uses `x` and `y` and this notation is used often by programmers for various points. `longitude` and `latitude` could be a better choice for a location, but I'd keep `x/y` to be consistent with the docs.

Comment (elimelec) [reply app/tests/LocationTest.kt:9]: One exception could be the usage output: `Usage: <program name> <user x coordinate> <user y coordinate> <shop data filename>.`

I think it would be easier to understand this message with `longitude/latitude` if you are not a developer.

File: app/src/CoffeeShopApp.kt (hunk 22 above)
Comment (elimelec) [reply app/src/CoffeeShopApp.kt:25]: I could ignore the bad rows from the csv file. The user could still use the app with the other available coffee shops.

File: app/tests/CoffeeShopTest.kt (hunk 4 above)
Comment (elimelec) [reply app/tests/CoffeeShopTest.kt:14]: Ok, I'll keep this in mind.

File: app/tests/CoffeeShopWithDistanceTest.kt (hunk 15 above)
Comment (elimelec) [reply app/tests/CoffeeShopWithDistanceTest.kt]: I agree.

File: app/tests/CoffeeShopWithDistanceTest.kt (hunk 12 above)
Comment (elimelec) [reply app/tests/CoffeeShopWithDistanceTest.kt:11]: I tried to write only enough test to fail.

File: app/tests/LocationTest.kt (hunk 1 above)
Comment (alexandru-calinoiu) [reply app/tests/LocationTest.kt:8]: No, actually I was thinking about the free equality operator that you also found out about later on in the pr.

File: app/src/CoffeeShopApp.kt (hunk 22 above)
Comment (alexandru-calinoiu) [reply app/src/CoffeeShopApp.kt:25]: Yeah, good point, also should consider to report the bad rows (all of them) I always find it annoying when apps keep feeding me input errors one by one.

**Comment by alexandru-calinoiu** [comment]
Thank you for your hard work, will followup shortly via email.
